Enable building and running on Mono/Mac

mhutch opened the pull request.

dnfclas commented:
Hi __@mhutch__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

mhutch commented on build.sh:
```diff
@@ -0,0 +1,6 @@
+CONFIG="${1:-Release}"
+SLNDIR="$( dirname ${0} )"
+
+msbuild WebJobs.Script.proj "/p:Configuration=${CONFIG};SolutionDir=${SLNDIR})"
```
Fixed, thanks :)

Therzok commented on src/WebJobs.Script.Cli/Helpers/PlatformHelper.cs:
```diff
@@ -0,0 +1,13 @@
+    internal static class PlatformHelper
+    {
+        public static bool IsMono { get; }  = Type.GetType("Mono.Runtime") != null;
+        public static bool IsWindows { get; } = IsWindows = Environment.OSVersion.Platform == PlatformID.Win32NT;
```
`IsWindows = IsWindows`?

mhutch replied:
Oops.

Therzok commented on src/WebJobs.Script.WebHost/WebScriptHostManager.cs:
```diff
@@ -139,7 +139,12 @@ public void Initialize()
                         {
+ //Mono does not yet support HostingEnvironment.QueueBackgroundWorkItem
+#if MONO
+                            Task.Run(() => RunAndBlock());
```
Shouldn't this be `Warmup(_webHostSettings)`?

mhutch replied:
Good catch :) Fixed!

Therzok commented on src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs:
```diff
@@ -163,24 +164,65 @@ private static ScriptHostConfiguration CreateScriptHostConfiguration(WebHostSett
+            }
 
-                scriptHostConfig.HostConfig.HostId = hostId.ToLowerInvariant();
+            //length canmot exceed 32
```
cannot

mhutch replied:
We need a spell checker :)

fabiocav commented on src/WebJobs.Script/WebJobs.Script.csproj:
```diff
@@ -15,7 +15,7 @@
     <CodeAnalysisRuleSet>..\src.ruleset</CodeAnalysisRuleSet>
     <CodeAnalysisCulture>en-US</CodeAnalysisCulture>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
Same as in previous projects... revert this.

fabiocav commented on src/WebJobs.Script.WebHost/WebScriptHostManager.cs:
```diff
@@ -153,7 +158,11 @@ public void Initialize()
                     if (!_webHostSettings.IsSelfHost)
                     {
+#if MONO
+                        Task.Run(() => RunAndBlock());
```
Maybe not super important for what it's intended, but this will behave differently since we do signal the cancellation token that is provided to the function (if bound) to allow the function to gracefully cleanup when the runtime signals a shutdown. 
I guess we could register to be notified of shutdown (assuming that's also available) and pass our own token that we signal ourselves when notified, but again, probably not worth it for now.
A comment here might be helpful, though. Thoughts?

fabiocav commented on src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs:
```diff
@@ -163,24 +164,65 @@ private static ScriptHostConfiguration CreateScriptHostConfiguration(WebHostSett
+        //cannot start or end with dash.
+        //cannot have consecutive dashes.
+        //max length 32.
+        private static string MakeValidHostID(string id)
```
nit: For consitency, rename to `MakeValidHostId`

fabiocav commented on src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs:
```diff
@@ -163,24 +164,65 @@ private static ScriptHostConfiguration CreateScriptHostConfiguration(WebHostSett
             // Otherwise, derive it from machine name and folder name
-            string hostId = _settingsManager.AzureWebsiteDefaultSubdomain
-                ?? $"{Environment.MachineName}-{Path.GetFileName(Environment.CurrentDirectory)}";
+            string hostId = MakeValidHostID(
```
We want to make sure that, on Azure, we use what is provided by the environment with no modification (and it should be safe, so the extra processing isn't needed), so we can probably change this to:
`_settingsManager.AzureWebsiteDefaultSubdomain ?? MakeValidHostID($"{Environment.MachineName}-{Path.GetFileName(Environment.CurrentDirectory)}");`

fabiocav commented on src/WebJobs.Script/Description/DotNet/FSharp/FSharpCompilationService.cs:
```diff
@@ -196,6 +203,8 @@ public ICompilation GetFunctionCompilation(FunctionMetadata functionMetadata)
             return new FSharpCompilation(errors, assemblyOption);
         }
 
+        static bool _isMono = Type.GetType("Mono.Runtime") != null;
```
We should probably move the `PlatformHelper` to `WebJobs.Script` and use it here as well, so we can remove this and have that logic centralized there.

mhutch commented on src/WebJobs.Script.Cli/WebJobs.Script.Cli.csproj:
```diff
@@ -15,7 +15,7 @@
     <CodeAnalysisRuleSet>..\src.ruleset</CodeAnalysisRuleSet>
     <CodeAnalysisCulture>en-US</CodeAnalysisCulture>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
There were a few warnings unrelated to my changes that caused the build to fail, and this was easier than fixing them: https://gist.github.com/mhutch/618f8d8f431ef950cfbf281beb16a12c

fabiocav replied:
I see. That being the case, feel free to leave as is. I'll take note and we'll have a chance to look at this again when merging this code into `dev`

mhutch commented:
I think I addressed all of the feedback. The fix for the warnings was pretty invasive though.

I don't have any plans to investigate node support in the near future.

fabiocav commented:
LGTM! 👍 

Would be good to have @ahmelsayed look at this as well. (he is currently on vacation)

We may also want to consider reverting 162703e as it was indeed more invasive and significantly increased the scope of the changes.

mhutch commented:
Well... I could have just ignored the warnings with pragmas but they were legitimate so I tried to actually fix them. Most of the change is that instead of duplicating the existing helpers that I needed to fix the warnings, I moved them to a place they could be shared. I also optimised TaskExtensions a bit while I was at it.

fabiocav commented:
Good point. Those were indeed good changes and actually addressed the issues causing the warning.

Just brought it up based on your previous comment so @ahmelsayed would be aware.

fabiocav commented:
@lindydonna FYI. Please look at the comment above. We need to make sure we have a work item to setup CI building and testing this on Mono on Linux, since that is not currently part of our workflow and it would be easy to introduce breaking changes if we don't have that in place.

lindydonna commented:
@fabiocav Agreed, plus afaik automation is also not fully in place for the CLI. (This will be part of the merge into master.)

dsyme commented:
Nice job @mhutch

ahmelsayed commented:
This has moved to https://github.com/Azure/azure-functions-cli/pull/19 and https://github.com/Azure/azure-webjobs-sdk-script/pull/1083
